csrc/scheduler/utils.cpp

```diff
@@ -2656,6 +2659,367 @@ void moveNonConcretizedBroadcastInnermost(
+  std::unique_ptr<IdModel> id_model_;
+  ValGroups ref_id_groups_;
+  ValGroups all_ancestors_of_ref_;
+};
+
+void LoopDomainScheduler::schedule(TensorView* tv) const {
+  // Quick shortcut
+  if (ref_id_groups_ == graph().toGroups(tv->getLoopDomain())) {
+    // No need to change the current loop domain
+    return;
+  }
+
+  // All of the existing IDs are reused as much as possible to
+  // minimize creating new IDs.
+  auto all_ids = tv->domain()->allIDs();
+  std::unordered_map<ValGroup, IterDomain*> group_to_id;
+  ValGroups all_id_groups;
+  for (auto id : all_ids) {
+    const auto& group = graph().toGroup(id);
+    group_to_id.emplace(group, id);
+    all_id_groups.pushBack(group);
+  }
+
+  // New loop domain to set for the tv
+  std::vector<IterDomain*> loop_domain(ref_loop_dom_.size());
+
+  // Find missing IDs.
+  bool has_missing_ids = false;
+  for (const auto i : c10::irange(ref_loop_dom_.size())) {
+    const auto& ref_id_group = ref_id_groups_.at((int64_t)i);
+    if (all_id_groups.has(ref_id_group)) {
+      // This loop ID already exists.
+      auto it = group_to_id.find(ref_id_group);
+      NVF_ERROR(it != group_to_id.end());
+      loop_domain.at(i) = it->second;
+    } else {
+      // Need to create a new ID for the loop ID
+      has_missing_ids = true;
+      // TODO: Don't force mapping at this point since that may not be necessary
+      auto clone = ref_loop_dom_.at(i)->cloneWithoutRFactor(true);
+      loop_domain.at(i) = clone;
+      group_to_id.emplace(ref_id_group, clone);
+      all_id_groups.pushBack(ref_id_group);
+    }
+  }
+
+  // If no new ID is created, no expr replay is necessary
+  if (!has_missing_ids) {
+    tv->setLoopDomain(loop_domain);
+    return;
+  }
+
+  const auto path_from_ref = getReplayPath(tv);
+  const ExprGroups all_existing_expr_groups =
+      graph().toGroups(tv->domain()->allExprs());
+
+  // Replay the path on the target tensor
+  for (const auto& [expr_g, dir] : path_from_ref) {
+    // Skip if the tensor already has the expr
+    if (all_existing_expr_groups.has(expr_g)) {
+      continue;
+    }
+
+    const auto input_groups = inputGroups(graph(), expr_g, dir);
+    const auto output_groups = outputGroups(graph(), expr_g, dir);
+
+    // All inputs must be already in all_id_groups
+    auto inputs_it = std::find_if(
+        input_groups.begin(),
+        input_groups.end(),
+        [&](const ValGroup& input_g) -> bool {
+          return !all_id_groups.has(input_g);
+        });
+    NVF_ERROR(
+        inputs_it == input_groups.end(),
+        "Unknown input group found: ",
+        nvfuser::toString(*inputs_it));
+
+    // Clone outputs if not found
+    for (const auto& output_g : output_groups) {
+      if (all_id_groups.has(output_g)) {
+        continue;
+      }
+
+      // No need to force exact mapping since this clone is going to
+      // be connected with tv
+      auto clone = output_g->front()->as<IterDomain>()->cloneWithoutRFactor();
```

Comment:
Should we use this?
https://github.com/NVIDIA/Fuser/blob/2980c3c25999bbb94d39a79044fcbf968fa78b1f/csrc/id_model/schedule.h#L14-L19

Reply:
That's a good idea. Then, this should never pick reduction IDs over non-reduction IDs.